server/raft_transport.go

```diff
@@ -70,11 +83,50 @@ func newRPCTransport(gossip *gossip.Gossip, rpcServer *rpc.Server, rpcContext *r
 			t.RaftMessage, &multiraft.RaftMessageRequest{}); err != nil {
 			return nil, err
 		}
+
+		if err := t.rpcServer.RegisterAsync(multiPackageMessageName, false, /*not public*/
+			t.MultiPackageMessage, &multiraft.MultiPackageMessageRequest{}); err != nil {
+			return nil, err
+		}
 	}
 
 	return t, nil
 }
 
+// MultiPackageMessage merges incoming multi-package request into the real request,
+// and proxies to the listening server interface.
+func (t *rpcTransport) MultiPackageMessage(args proto.Message, callback func(proto.Message, error)) {
+	req := args.(*multiraft.MultiPackageMessageRequest)
+
+	t.multiPackageMessageLock.Lock()
+	_, ok := t.multiPackageMessage[req.GroupID]
+	if !ok {
+		t.multiPackageMessage[req.GroupID] = &[]byte{}
+	}
+	data := t.multiPackageMessage[req.GroupID]
+	*data = append(*data, req.RawBytes...)
```

Comment:
We need to consider the possibility of a leader failing while it is sending the snapshot, and then a new leader restarting from packet 0 (among other potential failures). We need some sort of request ID so that we can avoid mixing packets from distinct streams (and potentially support multiple incoming multi-packet sequences per group).

This can be tricky, and it should get a lot easier with GRPC. Let's leave this part out of the change for now, and I'll try to dust off the GRPC prototype.